sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -147,31 +148,133 @@ class FrameLessOffsetWindowFunctionFrame(
-    inputIndex = offset
+  }
+
+  private val doWrite = if (ignoreNulls && offset > 0) {
+    // For illustration, here is one example: the input data contains six rows,
+    // and the input values of each row are: null, x, null, null, y, null, z, null.
+    // We use Lead(input, 2) with IGNORE NULLS and the process is as follows:
+    // 1. current row -> null, next selected row -> y, output: y;
+    // 2. current row -> x, next selected row -> z, output: z;
+    // 3. current row -> null, next selected row -> z, output: z;
+    // 4. current row -> null, next selected row -> z, output: z;
+    // 5. current row -> y, next selected row -> empty, output: null;
+    // ... next selected row is empty, all following return null.
+    (current: InternalRow) =>
+      if (current.isNullAt(idx)) {
+        if (nextSelectedRow == EmptyRow) {
+          // Use default values since the offset row whose input value is not null does not exist.
+          fillDefaultValue(current)
+        } else {
+          projection(nextSelectedRow)
+        }
+      } else {
+        skipNonNullCount -= 1
+        findNextRowWithNonNullInput
+        if (skipNonNullCount == offset) {
+          projection(nextSelectedRow)
+        } else {
+          // Use default values since the offset row whose input value is not null does not exist.
+          fillDefaultValue(current)
+          nextSelectedRow = EmptyRow
+        }
+      }
+  } else if (ignoreNulls && offset < 0) {
+    // For illustration, here is one example: the input data contains six rows,
+    // and the input values of each row are: null, x, null, null, y, null, z, null.
+    // We use Lag(input, 1) with IGNORE NULLS and the process is as follows:
+    // 1. current row -> null, next selected row -> empty, output: null;
+    // 2. current row -> x, next selected row -> empty, output: null;
+    // 3. current row -> null, next selected row -> x, output: x;
+    // 4. current row -> null, next selected row -> x, output: x;
+    // 5. current row -> y, next selected row -> x, output: x;
+    // 6. current row -> null, next selected row -> y, output: y;
+    // 7. current row -> z, next selected row -> y, output: y;
+    // 8. current row -> z, next selected row -> z, output: z;
+    val absOffset = Math.abs(offset)
+    (current: InternalRow) =>
+      if (nextSelectedRow == EmptyRow && skipNonNullCount == absOffset) {
+        do {
+          val r = WindowFunctionFrame.getNextOrNull(inputIterator)
+          if (!r.isNullAt(idx)) {
+            nextSelectedRow = r
+          }
+          inputIndex += 1
+        } while (nextSelectedRow == EmptyRow && inputIndex < input.length)
+      }
+      if (nextSelectedRow == EmptyRow) {
+        // Use default values since the offset row whose input value is not null does not exist.
+        fillDefaultValue(current)
+      } else {
+        projection(nextSelectedRow)
+      }
+      if (!current.isNullAt(idx)) {
+        if (skipNonNullCount < absOffset) {
```

Comment:
There is a bug that `skipNonNullCount` is never decreased. Please update the tests to capture this kind of bugs.

Reply:
Yes. Thank you

Reply:
The original idea is that as long as skipNonNullCount is increased to absOffset, it will no longer increase